src/rpc/mining.cpp

```diff
@@ -578,24 +585,36 @@ UniValue getblocktemplate(const UniValue& params, bool fHelp)
             nTransactionsUpdatedLastLP = nTransactionsUpdatedLast;
         }
 
-        // Release the wallet and main lock while waiting
-        LEAVE_CRITICAL_SECTION(cs_main);
         {
-            checktxtime = boost::get_system_time() + boost::posix_time::minutes(1);
+            checktxtime = boost::get_system_time() + boost::posix_time::seconds(10);
 
+            int nHeight = chainActive.Tip()->nHeight;
             boost::unique_lock<boost::mutex> lock(csBestBlock);
             while (chainActive.Tip()->GetBlockHash() == hashWatchedChain && IsRPCRunning())
             {
-                if (!cvBlockChange.timed_wait(lock, checktxtime))
-                {
-                    // Timeout: Check transactions for update
-                    if (mempool.GetTransactionsUpdated() != nTransactionsUpdatedLastLP)
-                        break;
-                    checktxtime += boost::posix_time::seconds(10);
+                // Before waiting, generate the coinbase for the block following the next
+                // block (since this is cpu-intensive), so that when next block arrives,
+                // we can quickly respond with the template for following block.
+                if (!next_coinbase_mtx && IsShieldedMinerAddress(minerAddress)) {
+                    next_coinbase_mtx = CreateCoinbaseTransaction(Params(), CAmount{0}, minerAddress, nHeight + 2);
                 }
+                // Release the main lock while waiting
+                LEAVE_CRITICAL_SECTION(cs_main);
+                bool timedout = !cvBlockChange.timed_wait(lock, checktxtime);
+                ENTER_CRITICAL_SECTION(cs_main);
+
+                checktxtime += boost::posix_time::seconds(10);
+                // Timeout: Check transactions for update
+                if (timedout && mempool.GetTransactionsUpdated() != nTransactionsUpdatedLastLP) {
+                    next_coinbase_mtx = std::nullopt;
```

Comment:
AFAICT, if an empty block is returned from `getblocktemplate`, this logic will ensure that we keep returning empty blocks until either the chain tip changes, or a new transaction arrives in the mempool. If we are going to generate empty blocks, it seems preferable to always generate a non-empty block if we generated an empty block on the previous call and then time out.

Is there an existing static variable that will be set precisely in this case, that could be used to disambiguate? (e.g. `pindexPrev = nullptr` being set below if we generated an empty block)

Reply:
I don't quite understand; take a look at the fairly large [comment](https://github.com/zcash/zcash/pull/4886/files#r539365809) I just wrote in reply to Nick and your comments, and let me know if that clears things up or if there is still some code change needed.

Are you concerned about multiple consecutive empty blocks (in the blockchain as opposed to just being returned from this RPC)? That can happen even without this PR if the blocks are solved close in time (because the mempool will be empty after the first block in the sequence).

Reply:
At this point in the code, `nTransactionsUpdatedLastLP` will be set to the value of `mempool.GetTransactionsUpdated()` in the previous call to `getblocktemplate`. The output of `mempool.GetTransactionsUpdated()` only changes in two situations:
- `UpdateTip` is called (in which case the tip hash also changes)
- A new transaction is added to the mempool.

If the previous call to `getblocktemplate` triggered an empty block, and the node _doesn't_ receive any new transactions in the intervening 10 seconds, then this logic will cause another empty block to be generated, ignoring any transactions currently in the mempool.